### ISSUE TITLE
Is there any way I can see the actual pipeline definition of Blog or Doc Recipes?

### ISSUE THREAD

**Issue by AllanXu49** [statiqdev/Statiq#415]
I understand that the Wyam theme source code is available in below repository to review and customize:

https://github.com/Wyamio/Wyam/tree/develop/themes

I am looking for a similar source for Wyam Recipes. My understanding is that Wyam Recipes are pipeline definitions like what we can have in the config file. Is there any way that I can see the actual Recipes pipeline definitions?

**Comment by daveaglick** [comment]
Absolutely -

Blog: https://github.com/Wyamio/Wyam/blob/develop/src/extensions/Wyam.Blog/Blog.cs
Docs: https://github.com/Wyamio/Wyam/blob/develop/src/extensions/Wyam.Docs/Docs.cs

The syntax looks slightly different since it's straight C# code without the benefit of the extra syntax sugar in the configuration file, but it's also one-to-one (I.e., that code could be represented completely in a config file with minimal changes).

**Comment by AllanXu49** [comment]
Thank you for help.

I was hoping that the Syntax is similar what we have in config files. That way it would be more straightforward to customize the out of the box receipts.

Is there any way to get the equivalent syntax for recipes that can be defined in the site's config file?

I like to make my site self-sufficient such that it includes its own theme as well as receipt defined in the config file. Or, is this a bad idea?

**Comment by daveaglick** [comment]
The translation isn't hard, in fact the code from these C# files should work as-is in the config files. All the extra syntax sugar in the config file is optional. The config file just adds implicit factory methods for each of the modules so you don't need to instantiate them with `new`. It also does some magic for delegates with the `@doc` syntax

**Comment by daveaglick** [comment]
Sorry, didn't mean to hit the close button...

**Comment by AllanXu49** [comment]
Thanks a lot David for answering all my questions. I appreciate it. 

I am trying to learn Wyam more in depth, then start using it.

Once I better learned Waym, I will start to blog about my experiences and probably create some Wyam tutorials for designers. I need to build my site 1st :)

I am confident that Wyam will be .NET flagship static site generator and I hope it ranks above other site generators since its design concepts are superior.

**Comment by daveaglick** [comment]
No problem, that's what I'm here for :). General questions like this also help other people in the future since the issue tracker becomes something of a knowledge base. Feel free to ask away.

**Comment by JonDouglas** [comment]
@daveaglick What is the preferred method of creating recipes? I'm currently working on an e-book recipe and I found a few roadblocks off extending the docs recipe as I noticed some crucial logic is actually done within some of the pipelines and  unless I change the default doc recipe, roll my own recipe, or override a pipeline(https://wyam.io/docs/concepts/recipes#pipelines) it seems a bit of a edge-case.

For example:

Should I go down the C# path? This seems reasonable, but I'm worried about how one sets up the Wyam environment from a dev standpoint. I wasn't able to find much documentation on that item in the sense of creating a `Wyam.exe` locally to use with the custom recipe code.

EX:

https://github.com/Wyamio/Wyam/blob/master/src/extensions/Wyam.Docs/Docs.cs

If I customized my own `IRecipe`, how would I go about debugging it?

Would it be more feasible to use the `Wyam.Common` nuget package and extend off that instead?

https://github.com/Wyamio/Wyam/blob/832d828ecbdfa0649a9b15e7dd039a4436b1fd31/src/core/Wyam.Common/Configuration/IRecipe.cs

Similar to https://wyam.io/docs/extensibility/writing-a-module

The other option seems to convert the C# to config file, but I'd love to see an example of that done for the existing recipes. Do those exist?

Anyway, was really asking for what your advice would be with regards to developing and debugging a recipe that I can at the end add a pull request/etc so the community can use it.

Thanks again!
